File: apps/desktop/src/api/hosted_rooms.ts
/**
 * M1.5 Phase 6: hosted-room API client.
 *
 * Surface for the Desktop renderer to call the gateway's
 * hosted-rooms REST endpoints. The endpoints themselves live in
 * `gateway/hosted_rooms.py` and are exposed on the existing
 * `ZELOOApi` (see `./client.ts`).
 *
 * Phase 6 scope: list + get + create + disband. The four
 * "primary surface" verbs that zeloo_cli's /room slash command
 * already exposed. The query hook (`useRoomMembership`) reads
 * `listRooms`; the chat Mount in Phase 3.5 will read it on
 * session focus and populate the membership store.
 *
 * The `members` payload is a JSON array of `{ member_id, kind,
 * display_name?, profile?, capabilities? }` objects. We do not
 * re-shape it on the client — the consumer (a future Phase 7
 * room roster panel) reads the same shape the gateway wrote.
 */
import { ZELOOApi } from './client'

export type RoomMember = {
  member_id: string
  kind: 'user' | 'agent' | 'sub-agent' | 'gateway' | 'system'
  display_name?: string
  profile?: string
  capabilities?: string[]
}

export interface RoomInfo {
  room_id: string
  name: string
  members: RoomMember[]
  authority_gateway_id: string
  authority_epoch: number
  next_seq: number
  revision: number
  created_at: number
  updated_at: number
  disbanded_at: number | null
}

export interface ListRoomsResponse {
  rooms: RoomInfo[]
  total: number
  limit: number
  offset: number
}

const LIST_TIMEOUT_MS = 10_000
const MUTATE_TIMEOUT_MS = 10_000

/** Phase 6: list active (non-disbanded) rooms. */
export async function listRooms(options: { limit?: number; offset?: number } = {}): Promise<ListRoomsResponse> {
  const limit = options.limit ?? 50
  const offset = options.offset ?? 0
  return ZELOOApi<ListRoomsResponse>({
    path: `/api/hosted_rooms?limit=${limit}&offset=${offset}`,
    timeoutMs: LIST_TIMEOUT_MS
  })
}

/** Phase 6: get one room by id (full state, including events). */
export async function getRoom(roomId: string): Promise<RoomInfo> {
  return ZELOOApi<RoomInfo>({
    path: `/api/hosted_rooms/${encodeURIComponent(roomId)}`,
    timeoutMs: LIST_TIMEOUT_MS
  })
}

/**
 * Phase 9: list one room's event log (oldest first). Bounded read
 * so a chatty agent run cannot pin a desktop renderer to a single
 * 100k-row query. The kinds allowlist is a future hook for an
 * event-type filter chip; today we read all kinds.
 */
export interface RoomEvent {
  room_id: string
  seq: number
  event_id: string
  kind: string
  actor_json?: string
  actor?: { actor_id: string; kind: string; display_name?: string }
  authority_epoch?: number
  payload_json?: string
  payload?: Record<string, unknown>
  created_at: number
}

export interface ListRoomEventsResponse {
  events: RoomEvent[]
  total: number
  limit: number
  offset: number
}

export async function listRoomEvents(
  roomId: string,
  options: { limit?: number; offset?: number; kinds?: string[] } = {}
): Promise<ListRoomEventsResponse> {
  const limit = options.limit ?? 100
  const offset = options.offset ?? 0
  let query = `limit=${limit}&offset=${offset}`
  if (options.kinds && options.kinds.length > 0) {
    query += `&kinds=${options.kinds.map(k => encodeURIComponent(k)).join(',')}`
  }
  const resp = await ZELOOApi<{ events: Array<Record<string, unknown>>; total: number; limit: number; offset: number }>({
    path: `/api/hosted_rooms/${encodeURIComponent(roomId)}/events?${query}`,
    timeoutMs: LIST_TIMEOUT_MS
  })
  return {
    events: (resp.events ?? []).map((row) => {
      const payload = typeof row.payload_json === 'string' ? safeParse(row.payload_json) : (row.payload as Record<string, unknown> | undefined)
      const actor = typeof row.actor_json === 'string' ? safeParse(row.actor_json) : (row.actor as { actor_id: string; kind: string; display_name?: string } | undefined)
      return {
        ...row,
        payload,
        actor
      } as RoomEvent
    }),
    total: resp.total,
    limit: resp.limit,
    offset: resp.offset
  }
}

function safeParse(raw: string): unknown {
  try {
    return JSON.parse(raw)
  } catch {
    return undefined
  }
}

/** Phase 6: create a new room. */
export async function createRoom(input: {
  name: string
  members: Array<Omit<RoomMember, 'kind'> & { kind: RoomMember['kind'] }>
}): Promise<RoomInfo> {
  return ZELOOApi<RoomInfo>({
    path: '/api/hosted_rooms',
    method: 'POST',
    body: input,
    timeoutMs: MUTATE_TIMEOUT_MS
  })
}

/** Phase 6: disband (soft-delete) a room. */
export async function disbandRoom(roomId: string): Promise<{ ok: boolean }> {
  return ZELOOApi<{ ok: boolean }>({
    path: `/api/hosted_rooms/${encodeURIComponent(roomId)}`,
    method: 'DELETE',
    timeoutMs: MUTATE_TIMEOUT_MS
  })
}

File: zeloo_cli/web_routers/hosted_rooms.py
"""
M1.5 Phase 6 part 3: hosted-rooms FastAPI router for ``Zeloo serve``.

Mounts ``/api/hosted_rooms`` on the same FastAPI app that already
serves ``/api/profiles``, ``/api/sessions``, and the other desktop
client surfaces on port 9119. The previous Phase 6 part 2 commit
added the same routes on the gateway api_server (aiohttp, port
8642) so the room CRUD surface works against the gateway
directly. THIS commit closes the loop on the desktop's real
path: ``Zeloo serve`` -> IPC ``Zeloo:api`` -> HTTP fetch on 9119
-> THIS router -> in-process call to ``gateway.hosted_rooms``
(same sqlite, same authority, same author/committer signature).

The two layers are deliberately not unified: the gateway layer
is the authoritative surface for messaging / multi-tenant
deployments, the serve layer is the desktop-local surface that
forks no new process. When a desktop renderer runs the M1.5
``/api/hosted_rooms`` query through its standard ``ZELOOApi``
helper, this router answers.
"""
from __future__ import annotations

import json
import logging
import time
from typing import Any, List, Optional

from fastapi import APIRouter, HTTPException, Query
from pydantic import BaseModel, Field

from gateway import hosted_rooms

router = APIRouter()
_log = logging.getLogger("zeloo_cli.web_routers.hosted_rooms")


# ---- request / response models ---------------------------------------------

class RoomMemberIn(BaseModel):
    member_id: str
    kind: str  # 'user' | 'agent' | 'sub-agent' | 'gateway' | 'system'
    display_name: Optional[str] = None
    profile: Optional[str] = None
    capabilities: Optional[List[str]] = None


class CreateRoomIn(BaseModel):
    name: str = Field(..., min_length=1, max_length=200)
    members: List[RoomMemberIn] = Field(..., min_length=1)
    room_id: Optional[str] = None  # server mints if absent


class CreateRoomOut(BaseModel):
    room: dict


# ---- helpers ----------------------------------------------------------------

def _serialize_room(room: dict) -> dict:
    members = room.get("members")
    if isinstance(members, str):
        try:
            members = json.loads(members)
        except json.JSONDecodeError:
            members = []
    elif members is None:
        members = []
    return {
        "room_id": room.get("room_id"),
        "name": room.get("name"),
        "members": members,
        "authority_gateway_id": room.get("authority_gateway_id"),
        "authority_epoch": room.get("authority_epoch", 0),
        "next_seq": room.get("next_seq", 0),
        "revision": room.get("revision", 0),
        "created_at": room.get("created_at", 0),
        "updated_at": room.get("updated_at", 0),
        "disbanded_at": room.get("disbanded_at"),
    }


def _serialize_event(ev: dict) -> dict:
    """Decode payload_json / actor_json into nested objects so the
    desktop typed client sees the same shape the gateway wrote."""
    row = dict(ev)
    pj = row.pop("payload_json", None)
    aj = row.pop("actor_json", None)
    if isinstance(pj, str):
        try:
            row["payload"] = json.loads(pj)
        except json.JSONDecodeError:
            row["payload"] = None
    if isinstance(aj, str):
        try:
            row["actor"] = json.loads(aj)
        except json.JSONDecodeError:
            row["actor"] = None
    return row


# ---- routes -----------------------------------------------------------------

@router.get("/api/hosted_rooms")
def list_rooms(
    limit: int = Query(50, ge=1, le=500),
    offset: int = Query(0, ge=0),
    include_disbanded: bool = Query(False),
):
    rooms = hosted_rooms.list_rooms(
        hosted_rooms.default_db_path(),
        include_disbanded=include_disbanded,
        limit=limit,
        offset=offset,
    )
    return {
        "rooms": [_serialize_room(r) for r in rooms],
        "total": len(rooms),
        "limit": limit,
        "offset": offset,
    }


@router.post("/api/hosted_rooms", status_code=201)
def create_room(body: CreateRoomIn):
    # The desktop caller may either let the server mint a room_id
    # (the common case) or supply one explicitly. The gateway's
    # create_room requires room_id as a keyword, so we generate one
    # server-side when the body omits it.
    room_id = body.room_id
    if not room_id:
        ts_hex = format(int(time.time() * 1_000_000) & 0xFFFFFFFF, "08x")
        room_id = f"r-{hosted_rooms.local_authority_gateway_id()[:8]}-{ts_hex}"
    try:
        result = hosted_rooms.create_room(
            hosted_rooms.default_db_path(),
            room_id=room_id,
            name=body.name,
            members=[m.model_dump() for m in body.members],
            authority_gateway_id=hosted_rooms.local_authority_gateway_id(),
        )
    except (ValueError, TypeError) as exc:
        raise HTTPException(status_code=400, detail=str(exc))
    return {"room": _serialize_room(result)}


@router.get("/api/hosted_rooms/{room_id}")
def get_room(room_id: str):
    rooms = hosted_rooms.list_rooms(
        hosted_rooms.default_db_path(),
        include_disbanded=True,
        limit=500,
    )
    match = next((r for r in rooms if r.get("room_id") == room_id), None)
    if match is None:
        raise HTTPException(status_code=404, detail="room not found")
    return _serialize_room(match)


@router.delete("/api/hosted_rooms/{room_id}")
def disband_room(room_id: str):
    rooms = hosted_rooms.list_rooms(
        hosted_rooms.default_db_path(),
        include_disbanded=True,
        limit=500,
    )
    match = next((r for r in rooms if r.get("room_id") == room_id), None)
    if match is None:
        raise HTTPException(status_code=404, detail="room not found")
    try:
        result = hosted_rooms.disband_room(
            hosted_rooms.default_db_path(),
            room_id=room_id,
            expected_gateway_id=match.get("authority_gateway_id", ""),
            expected_epoch=match.get("authority_epoch", 0),
        )
    except (ValueError, TypeError) as exc:
        raise HTTPException(status_code=400, detail=str(exc))
    if result is None:
        raise HTTPException(status_code=404, detail="room not found")
    return {"ok": True, "room": _serialize_room(result)}


@router.get("/api/hosted_rooms/{room_id}/events")
def list_room_events(
    room_id: str,
    limit: int = Query(100, ge=1, le=500),
    offset: int = Query(0, ge=0),
    kinds: Optional[str] = Query(None, description="comma-separated kind allowlist"),
):
    kind_list = [k.strip() for k in (kinds or "").split(",") if k.strip()] or None
    try:
        events = hosted_rooms.list_room_events(
            hosted_rooms.default_db_path(),
            room_id=room_id,
            limit=limit,
            offset=offset,
            kinds=kind_list,
        )
    except (ValueError, TypeError) as exc:
        raise HTTPException(status_code=400, detail=str(exc))
    return {
        "events": [_serialize_event(e) for e in events],
        "total": len(events),
        "limit": limit,
        "offset": offset,
    }
